File: app/src/main/java/com/wafflestudio/snutt_staging/manager/NotiManager.java
package com.wafflestudio.snutt_staging.manager;

import android.util.Log;

import com.google.common.base.Preconditions;
import com.wafflestudio.snutt_staging.SNUTTApplication;
import com.wafflestudio.snutt_staging.model.Lecture;
import com.wafflestudio.snutt_staging.model.Notification;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

/**
 * Created by makesource on 2017. 2. 27..
 */

public class NotiManager {
    private static final String TAG = "NOTIFICATION_MANAGER" ;

    private static NotiManager singleton;
    private SNUTTApplication app;
    private List<Notification> notifications;
    private boolean fetched;

    private NotiManager(SNUTTApplication app) {
        Preconditions.checkNotNull(app);
        this.app = app;
        this.notifications = new ArrayList<>();
        this.fetched = false;
    }

    public static NotiManager getInstance(SNUTTApplication app) {
        singleton = new NotiManager(app);
        return singleton;
    }

    public static NotiManager getInstance() {
        return singleton;
    }

    public List<Notification> getNotifications() {
        return notifications;
    }

    public void loadData(int offset, final Callback callback) {
        String token = PrefManager.getInstance().getPrefKeyXAccessToken();
        Map query = new HashMap();

        query.put("limit", 20);
        query.put("offset", offset);
        app.getRestService().getNotification(token, query, new Callback<List<Notification>>() {
            @Override
            public void success(List<Notification> notificationList, Response response) {
                Log.d(TAG, "get notification success!");
                removeProgressBar();
                for (Notification notification : notificationList) {
                    notifications.add(notification);
                }
                if (callback != null) callback.success(notificationList, response);
            }

            @Override
            public void failure(RetrofitError error) {
                Log.e(TAG, "get notification failed.");
                if (callback != null) callback.failure(error);
            }
        });
    }

    public void addProgressBar() {
        notifications.add(null);
    }

    public void removeProgressBar() {
        notifications.remove(notifications.size() - 1);
    }

    public boolean getFetched() {
        return fetched;
    }

    public void setFetched(boolean fetched) {
        this.fetched = fetched;
    }

    public void refreshNotification(final Callback callback) {
        notifications.clear();
        String token = PrefManager.getInstance().getPrefKeyXAccessToken();
        Map query = new HashMap();
        query.put("limit", 20);
        query.put("offset", 0);
        app.getRestService().getNotification(token, query, new Callback<List<Notification>>() {
            @Override
            public void success(List<Notification> notificationList, Response response) {
                Log.d(TAG, "get notification success!");
                for (Notification notification : notificationList) {
                    notifications.add(notification);
                }
                if (callback != null) callback.success(notificationList, response);
            }

            @Override
            public void failure(RetrofitError error) {
                Log.e(TAG, "get notification failed.");
                if (callback != null) callback.failure(error);
            }
        });
    }

}

File: app/src/main/java/com/wafflestudio/snutt2/manager/NotiManager.kt
package com.wafflestudio.snutt2.manager

import android.util.Log
import com.wafflestudio.snutt2.network.dto.core.NotificationDto
import com.wafflestudio.snutt2.network.SNUTTRestApi
import com.wafflestudio.snutt2.network.dto.GetNotificationCountResults
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by makesource on 2017. 2. 27..
 */
@Singleton
class NotiManager @Inject constructor(
    private val snuttRestApi: SNUTTRestApi,
    private val prefStorage: PrefStorage
) {
    private var notifications: MutableList<NotificationDto?> = ArrayList()
    var fetched: Boolean = false

    interface OnNotificationReceivedListener {
        fun notifyNotificationReceived()
    }

    private val listeners: MutableList<OnNotificationReceivedListener> = ArrayList()
    fun addListener(current: OnNotificationReceivedListener) {
        for (listener in listeners) {
            if (listener == current) {
                Log.w(TAG, "listener reference is duplicated !!")
                return
            }
        }
        listeners.add(current)
    }

    fun removeListener(current: OnNotificationReceivedListener) {
        for (listener in listeners) {
            if (listener == current) {
                listeners.remove(listener)
                break
            }
        }
        Log.d(TAG, "listener count: " + listeners.size)
    }

    /* local method */
    fun reset() {
        notifications = ArrayList()
        fetched = false
    }

    fun getNotifications(): List<NotificationDto?> {
        return notifications
    }

    fun hasNotifications(): Boolean {
        return if (notifications == null || notifications!!.size > 0) true else false
    }

    fun loadData(offset: Int): Single<List<NotificationDto>> {
        val token = prefStorage.prefKeyXAccessToken
        return snuttRestApi.getNotification(
            token!!,
            limit = 20,
            offset = offset.toLong(),
            explicit = 1
        )
            .subscribeOn(Schedulers.io())
            .doOnSuccess {
                Log.d(TAG, "get notification success!")
//                Refactoring FIXME: dirty progress
//                removeProgressBar()
                for (notification in it) {
                    notifications!!.add(notification)
                }
            }
            .doOnError {
                Log.e(TAG, "get notification failed.")
            }
    }

//    fun addProgressBar() {
//        // Refactoring FIXME: 더러운 코드
//        notifications!!.add(null)
//    }

//    fun removeProgressBar() {
//        notifications!!.removeAt(notifications!!.size - 1)
//    }

    fun refreshNotification(): Single<List<NotificationDto>> {
        val token = prefStorage.prefKeyXAccessToken
        return snuttRestApi.getNotification(
            token!!,
            limit = 20,
            offset = 0,
            explicit = 1
        )
            .subscribeOn(Schedulers.io())
            .doOnSuccess {
                Log.d(TAG, "get notification success!")
                notifications.clear()
                notifications.addAll(it)
            }
            .doOnError {
                Log.e(TAG, "get notification failed.")
            }
    }

    fun getNotificationCount(): Single<GetNotificationCountResults> {
        val token = prefStorage.prefKeyXAccessToken
        return snuttRestApi.getNotificationCount(token!!)
            .subscribeOn(Schedulers.io())
            .doOnSuccess {
                Log.d(TAG, "get notification count success!")
            }
            .doOnError {
                Log.e(TAG, "get notification count failed.")
            }
    }

    fun notifyNotificationReceived() {
        fetched = false
        for (listener in listeners) {
            listener.notifyNotificationReceived()
        }
    }

    companion object {
        private const val TAG = "NOTIFICATION_MANAGER"
    }
}
